README.md
# Terraformer ArcGIS JSON Parser

[![Build Status](https://travis-ci.org/Esri/terraformer-arcgis-parser.svg?branch=master)](https://travis-ci.org/Esri/terraformer-arcgis-parser)

> Two way conversion between [GeoJSON](http://geojson.org/geojson-spec.html) and [ArcGIS Geometry](http://help.arcgis.com/en/arcgisserver/10.0/apis/rest/geometry.html). Part of the [Terraformer](http://terraformer.io) project.

## Installing

### Node.js

    $ npm install terraformer-arcgis-parser

### Browser

In the browser, [Terraformer](http://github.com/esri/terraformer) is required.

## Documentation

For full documentation check out the [offical website](http://terraformer.io/arcgis-parser/).

### Node.js
```js
var ArcGIS = require('terraformer-arcgis-parser');

// parse ArcGIS JSON, convert it to a Terraformer.Primitive (GeoJSON)
var primitive = ArcGIS.parse({
    'x':-122.6764,
    'y':45.5165,
    'spatialReference': {
      'wkid': 4326
    }
  });

// take a Terraformer.Primitive or GeoJSON and convert it back to ArcGIS JSON
var point = ArcGIS.convert({
  'type': "Point",
  'coordinates': [45.5165, -122.6764]
});
```
### Browser
```html
  <!-- Load the main Terraformer library -->
  <script src="https://unpkg.com/terraformer/terraformer.js" type="text/javascript"></script>

  <!-- Load the ArcGIS Parser -->
  <script src="https://unpkg.com/terraformer-arcgis-parser/terraformer-arcgis-parser.js" type="text/javascript"></script>

  <!-- Use it! -->
  <script>
    var primitive = Terraformer.ArcGIS.parse({
      'x':-122.6764,
      'y':45.5165,
      'spatialReference': {
        'wkid': 4326
      }
    });

    // take a Terraformer.Primitive or GeoJSON and convert it to ArcGIS JSON
    var point = Terraformer.ArcGIS.convert({
      'type': "Point",
      'coordinates': [45.5165, -122.6764]
    });
  </script>
```
### TypeScript
```
import arcgis = require("terraformer-arcgis-parser");
```

## Resources

* [Terraformer Website](http://terraformer.io)
* [twitter@EsriPDX](http://twitter.com/esripdx)

## Issues

Find a bug or want to request a new feature?  Please let us know by submitting an issue.

## Contributing

Esri welcomes contributions from anyone and everyone. Please see our [guidelines for contributing](https://github.com/esri/contributing).

## Licensing

Copyright &copy; 2013-2018 Esri

A copy of the license is available in the repository's [LICENSE](./LICENSE) file.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [unreleased]

## [1.1.0] - 2018-07-09

### Added

* ArcGIS Extents/Envelopes are now converted to GeoJSON polygons 🙏CorinChappy🙏
* great new command line demo app 🙏gseyffert🙏
* an appropriate CRS is appended to output GeoJSON when web mercator geometries are encountered.

### Changed

* Ensure ring-order of GeoJSON is compliant with RFC 7946
* web mercator northings/eastings are no longer reprojected during conversion.
* TypeScript support files updated 🙏JeffJacobson🙏

## [1.0.5] - 2016-08-16

### Fixed
* Check for GeoJSON feature `id`s before assigning to output [#24](https://github.com/Esri/terraformer-arcgis-parser/pull/24)
* correct conversion of polygons with outer rings not containing holes [#28](https://github.com/Esri/terraformer-arcgis-parser/pull/28)

### Added
* typings for TypeScript folks (thx [@JeffJacobson](https://github.com/JeffJacobson)) [#34](https://github.com/Esri/terraformer-arcgis-parser/pull/34)

## [1.0.4] - 2014-06-17

### Fixed
* Account for breaking change in `Terraformer`

### Added
* support for `z` and `m` conversion

## [1.0.3] - 2015-02-24

### Fixed
* valid output on both ends of conversion [#19](https://github.com/Esri/terraformer-arcgis-parser/issues/19)

## [1.0.2] - 2014-02-10

### Added
* `parseCompressedGeometry()` [#10](https://github.com/Esri/terraformer-arcgis-parser/issues/10)

### Fixed
* `parse()` and `convert()` now close polygons during conversion. [#9](https://github.com/Esri/terraformer-arcgis-parser/issues/9)
* `parse()` now handles `compressedGeometry`

## [1.0.1] - 2013-12-04

### Fixed
* `ReferenceError: sr is not defined` in `ArcGIS.convert()`

## [1.0.0] - 2013-11-12

Initial Release.  Also available on NPM and Bower

```
npm install terraformer-arcgis-parser
bower install terraformer-arcgis-parser
```

[unreleased]: https://github.com/Esri/terraformer-arcgis-parser/compare/v1.1.0...HEAD
[1.1.0]: https://github.com/Esri/terraformer-arcgis-parser/compare/v1.0.5...v1.1.0
[1.0.5]: https://github.com/Esri/terraformer-arcgis-parser/compare/v1.0.4...v1.0.5
[1.0.4]: https://github.com/Esri/terraformer-arcgis-parser/compare/v1.0.3...v1.0.4
[1.0.3]: https://github.com/Esri/terraformer-arcgis-parser/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/Esri/terraformer-arcgis-parser/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/Esri/terraformer-arcgis-parser/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/Esri/terraformer-arcgis-parser/releases/tag/v1.0.0


test.ts
import * as TerraformerArcGIS from "./index";

console.assert(typeof TerraformerArcGIS !== undefined);

// parse an ArcGIS Geometry to GeoJSON
let geojsonPoint = TerraformerArcGIS.parse({
  x: -122.6764,
  y: 45.5165,
  spatialReference: {
    wkid: 4326
  }
});

// convert a GeoJSON object into an ArcGIS geometry
let arcgisPoint = TerraformerArcGIS.convert(geojsonPoint);


index.d.ts
/// <reference types="geojson" />

export interface SpatialReference {
}

export interface SpatialReferenceWkid extends SpatialReference {
    wkid?: number;
    latestWkid?: number;
    vcsWkid?: number;
    latestVcsWkid?: number;
}

export interface SpatialReferenceWkt extends SpatialReference {
    wkt?: string;
    latestWkt?: string;
}

export interface Geometry {
    spatialReference?: SpatialReference;
}

export interface HasZM {
    hasZ?: boolean;
    hasM?: boolean;
}

export interface Point extends Geometry {
    x: number;
    y: number;
    z?: number;
    m?: number;
}

export interface Polyline extends HasZM, Geometry {
    paths: number[][][];
}

export interface Polygon extends HasZM, Geometry {
    rings: number[][][];
}

export interface Multipoint extends HasZM, Geometry {
    points: number[][];
}

export interface Envelope extends Geometry {
    xmin: number;
    xmax: number;
    ymin: number;
    ymax: number;

    zmin?: number;
    zmax?: number;

    mmin?: number;
    mmax?: number;
}

export interface Feature {
    geometry: Geometry;
    attributes: any;
}

export interface Field {
    name: string;
    type: string;
    alias?: string;
    length?: number;
}

export type esriGeometryType = "esriGeometryPoint" | "esriGeometryMultipoint" | "esriGeometryPolyline" |
                               "esriGeometryPolygon" | "esriGeometryEnvelope";

export interface FeatureSet extends HasZM {
    objectIdFieldName?: string; // optional
    globalIdFieldName?: string; // optional
    displayFieldName?: string; // optional
    geometryType?: esriGeometryType; // for feature layers only
    spatialReference?: SpatialReference; // for feature layers only.
    fields?: Field[];
    features: Feature[];
}

export interface ParseOptions {
    sr?: number;
    idAttribute?: string;
}
export interface ConvertOptions {
    idAttribute?: string;
}

export function parse<T extends Geometry>(json: T, options?: ParseOptions): GeoJSON.GeometryObject;
export function parse<T extends GeoJSON.GeometryObject>(json: Feature, options?: ParseOptions): GeoJSON.Feature<T>;

export function convert<T extends GeoJSON.GeometryObject>(geoJSON: GeoJSON.FeatureCollection<T>, options?: ConvertOptions): FeatureSet;
export function convert<T extends GeoJSON.GeometryObject>(geoJSON: GeoJSON.Feature<T>, options?: ConvertOptions): Feature;
export function convert<T extends GeoJSON.GeometryObject>(geoJSON: T, options?: ConvertOptions): Geometry;


release.sh
#!/bin/bash

# config
VERSION=$(node --eval "console.log(require('./package.json').version);")
NAME=$(node --eval "console.log(require('./package.json').name);")

# build and test
npm test || exit 1

# checkout temp branch for release
git checkout -b gh-release

# create built library (and versioned copy)
grunt uglify

# force add files
git add terraformer-arcgis-parser.min.js -f

# commit changes with a versioned commit message
git commit -m "build $VERSION"

# push commit so it exists on GitHub when we run gh-release
git push https://github.com/Esri/terraformer-arcgis-parser gh-release

# create copy of minified file with version number appended
cp terraformer-arcgis-parser.min.js $NAME-$VERSION.min.js

# run gh-release to create the tag and push release to github
gh-release --assets $NAME-$VERSION.min.js

# remove copy after the asset is attached to the github release
rm $NAME-$VERSION.min.js

# checkout master and delete release branch locally and on GitHub
git checkout master
git branch -D gh-release
git push https://github.com/Esri/terraformer-arcgis-parser :gh-release

# publish release on NPM
npm publish

Gruntfile.js
var fs = require('fs');

module.exports = function (grunt) {
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),

    meta: {
      banner: '/*! Terraformer ArcGIS Parser - <%= pkg.version %> - <%= grunt.template.today("yyyy-mm-dd") %>\n' +
        '*   https://github.com/esri/terraformer-arcgis-parser\n' +
        '*   Copyright (c) 2013-<%= grunt.template.today("yyyy") %> Esri, Inc.\n' +
        '*   Licensed MIT */'
    },

    uglify: {
      options: {
        report: 'gzip',
        banner: '<%= meta.banner %>'
      },
      arcgis: {
        src: ["terraformer-arcgis-parser.js"],
        dest: 'terraformer-arcgis-parser.min.js'
      }
    },

    jasmine: {
      coverage: {
        src: [
          "terraformer-arcgis-parser.js"
        ],
        options: {
          specs: 'spec/*Spec.js',
          helpers:[
            "node_modules/terraformer/terraformer.js"
          ],
          //keepRunner: true,
          outfile: 'SpecRunner.html',
          // template: require('grunt-template-jasmine-istanbul'),
          // templateOptions: {
          //   coverage: './coverage/coverage.json',
          //   report: './coverage',
          //   thresholds: {
          //     lines: 80,
          //     statements: 80,
          //     branches: 75,
          //     functions: 80
          //   }
          // }
        }
      }
    },

    complexity: {
      generic: {
        src: [ 'terraformer-arcgis-parser.js' ],
        options: {
          jsLintXML: 'complexity.xml', // create XML JSLint-like report
          errorsOnly: false, // show only maintainability errors
          cyclomatic: 6,
          halstead: 15,
          maintainability: 65
        }
      }
    }
  });

  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-contrib-jasmine');
  grunt.loadNpmTasks('grunt-complexity');

  grunt.registerTask('test', ['jasmine']);
  grunt.registerTask('version', ['test', 'uglify']);
  grunt.registerTask('default', ['test']);
};

terraformer-arcgis-parser.js
/* globals Terraformer */
(function (root, factory) {

  // Node.
  if(typeof module === 'object' && typeof module.exports === 'object') {
    exports = module.exports = factory(require('terraformer'));
  }

  // Browser Global.
  if(typeof root.navigator === "object") {
    if (!root.Terraformer){
      throw new Error("Terraformer.ArcGIS requires the core Terraformer library. https://github.com/esri/Terraformer");
    }
    root.Terraformer.ArcGIS = factory(root.Terraformer);
  }

}(this, function(Terraformer) {
  var exports = {};

  // https://github.com/Esri/terraformer-arcgis-parser/issues/10
  function decompressGeometry(str) {
    var xDiffPrev = 0;
    var yDiffPrev = 0;
    var points = [];
    var x, y;
    var strings;
    var coefficient;

    // Split the string into an array on the + and - characters
    strings = str.match(/((\+|\-)[^\+\-]+)/g);

    // The first value is the coefficient in base 32
    coefficient = parseInt(strings[0], 32);

    for (var j = 1; j < strings.length; j += 2) {
      // j is the offset for the x value
      // Convert the value from base 32 and add the previous x value
      x = (parseInt(strings[j], 32) + xDiffPrev);
      xDiffPrev = x;

      // j+1 is the offset for the y value
      // Convert the value from base 32 and add the previous y value
      y = (parseInt(strings[j + 1], 32) + yDiffPrev);
      yDiffPrev = y;

      points.push([x / coefficient, y / coefficient]);
    }

    return points;
  }

  // checks if the first and last points of a ring are equal and closes the ring
  function closeRing(coordinates) {
    if (!pointsEqual(coordinates[0], coordinates[coordinates.length - 1])) {
      coordinates.push(coordinates[0]);
    }
    return coordinates;
  }

  // checks if 2 x,y points are equal
  function pointsEqual(a, b) {
    for (var i = 0; i < a.length; i++) {
      if (a[i] !== b[i]) {
        return false;
      }
    }
    return true;
  }

  // shallow object clone for feature properties and attributes
  // from http://jsperf.com/cloning-an-object/2
  function clone(obj) {
    var target = {};
    for (var i in obj) {
      if (obj.hasOwnProperty(i)) {
        target[i] = obj[i];
      }
    }
    return target;
  }

  // determine if polygon ring coordinates are clockwise. clockwise signifies outer ring, counter-clockwise an inner ring
  // or hole. this logic was found at http://stackoverflow.com/questions/1165647/how-to-determine-if-a-list-of-polygon-
  // points-are-in-clockwise-order
  function ringIsClockwise(ringToTest) {
    var total = 0,i = 0;
    var rLength = ringToTest.length;
    var pt1 = ringToTest[i];
    var pt2;
    for (i; i < rLength - 1; i++) {
      pt2 = ringToTest[i + 1];
      total += (pt2[0] - pt1[0]) * (pt2[1] + pt1[1]);
      pt1 = pt2;
    }
    return (total >= 0);
  }

  // This function ensures that rings are oriented in the right directions
  // outer rings are clockwise, holes are counterclockwise
  function orientRings(poly){
    var output = [];
    var polygon = poly.slice(0);
    var outerRing = closeRing(polygon.shift().slice(0));
    if(outerRing.length >= 4){
      if(!ringIsClockwise(outerRing)){
        outerRing.reverse();
      }

      output.push(outerRing);

      for (var i = 0; i < polygon.length; i++) {
        var hole = closeRing(polygon[i].slice(0));
        if(hole.length >= 4){
          if(ringIsClockwise(hole)){
            hole.reverse();
          }
          output.push(hole);
        }
      }
    }

    return output;
  }

  // This function flattens holes in multipolygons to one array of polygons
  // [
  //   [
  //     [ array of outer coordinates ]
  //     [ hole coordinates ]
  //     [ hole coordinates ]
  //   ],
  //   [
  //     [ array of outer coordinates ]
  //     [ hole coordinates ]
  //     [ hole coordinates ]
  //   ],
  // ]
  // becomes
  // [
  //   [ array of outer coordinates ]
  //   [ hole coordinates ]
  //   [ hole coordinates ]
  //   [ array of outer coordinates ]
  //   [ hole coordinates ]
  //   [ hole coordinates ]
  // ]
  function flattenMultiPolygonRings(rings){
    var output = [];
    for (var i = 0; i < rings.length; i++) {
      var polygon = orientRings(rings[i]);
      for (var x = polygon.length - 1; x >= 0; x--) {
        var ring = polygon[x].slice(0);
        output.push(ring);
      }
    }
    return output;
  }

  function coordinatesContainCoordinates(outer, inner){
    var intersects = Terraformer.Tools.arraysIntersectArrays(outer, inner);
    var contains = Terraformer.Tools.coordinatesContainPoint(outer, inner[0]);
    if(!intersects && contains){
      return true;
    }
    return false;
  }

  // do any polygons in this array contain any other polygons in this array?
  // used for checking for holes in arcgis rings
  function convertRingsToGeoJSON(rings){
    var outerRings = [];
    var holes = [];
    var x; // iterator
    var outerRing; // current outer ring being evaluated
    var hole; // current hole being evaluated

    // for each ring
    for (var r = 0; r < rings.length; r++) {
      var ring = closeRing(rings[r].slice(0));
      if(ring.length < 4){
        continue;
      }
      // is this ring an outer ring? is it clockwise?
      if(ringIsClockwise(ring)){
        var polygon = [ ring.slice().reverse() ]; // wind outer rings counterclockwise for RFC 7946 compliance
        outerRings.push(polygon); // push to outer rings
      } else {
        holes.push(ring.slice().reverse()); // wind inner rings clockwise for RFC 7946 compliance
      }
    }

    var uncontainedHoles = [];

    // while there are holes left...
    while(holes.length){
      // pop a hole off out stack
      hole = holes.pop();

      // loop over all outer rings and see if they contain our hole.
      var contained = false;
      for (x = outerRings.length - 1; x >= 0; x--) {
        outerRing = outerRings[x][0];
        if(coordinatesContainCoordinates(outerRing, hole)){
          // the hole is contained push it into our polygon
          outerRings[x].push(hole);
          contained = true;
          break;
        }
      }

      // ring is not contained in any outer ring
      // sometimes this happens https://github.com/Esri/esri-leaflet/issues/320
      if(!contained){
        uncontainedHoles.push(hole);
      }
    }

    // if we couldn't match any holes using contains we can now try intersects...
    while(uncontainedHoles.length){
      // pop a hole off out stack
      hole = uncontainedHoles.pop();

      // loop over all outer rings and see if any intersect our hole.
      var intersects = false;
      for (x = outerRings.length - 1; x >= 0; x--) {
        outerRing = outerRings[x][0];
        if(Terraformer.Tools.arraysIntersectArrays(outerRing, hole)){
          // the hole intersects the outer ring push it into our polygon
          outerRings[x].push(hole);
          intersects = true;
          break;
        }
      }

      // hole does not intersect ANY outer ring at this point
      // make it an outer ring.
      if(!intersects) {
        outerRings.push([hole.reverse()]);
      }
    }

    if(outerRings.length === 1){
      return {
        type: 'Polygon',
        coordinates: outerRings[0]
      };
    } else {
      return {
        type: 'MultiPolygon',
        coordinates: outerRings
      };
    }
  }

  // ArcGIS -> GeoJSON
  function parse(arcgis, options){
    var geojson = {};

    options = options || {};
    options.idAttribute = options.idAttribute || undefined;

    if (arcgis.spatialReference && (arcgis.spatialReference.wkid === 3857 || arcgis.spatialReference.wkid === 102100)) {
      geojson.crs = Terraformer.MercatorCRS;
    }

    if(typeof arcgis.x === 'number' && typeof arcgis.y === 'number'){
      geojson.type = "Point";
      geojson.coordinates = [arcgis.x, arcgis.y];
      if (arcgis.z || arcgis.m){
        geojson.coordinates.push(arcgis.z);
      }
      if (arcgis.m){
        geojson.coordinates.push(arcgis.m);
      }
    }

    if(arcgis.points){
      geojson.type = "MultiPoint";
      geojson.coordinates = arcgis.points.slice(0);
    }

    if(arcgis.paths) {
      if(arcgis.paths.length === 1){
        geojson.type = "LineString";
        geojson.coordinates = arcgis.paths[0].slice(0);
      } else {
        geojson.type = "MultiLineString";
        geojson.coordinates = arcgis.paths.slice(0);
      }
    }

    if(arcgis.rings) {
      geojson = convertRingsToGeoJSON(arcgis.rings.slice(0));
    }

    if(
      typeof arcgis.xmin === "number" &&
      typeof arcgis.ymin === "number" &&
      typeof arcgis.xmax === "number" &&
      typeof arcgis.ymax === "number"
    ) {
      geojson.type = "Polygon";
      geojson.coordinates = [[
        [arcgis.xmax, arcgis.ymax],
        [arcgis.xmin, arcgis.ymax],
        [arcgis.xmin, arcgis.ymin],
        [arcgis.xmax, arcgis.ymin],
        [arcgis.xmax, arcgis.ymax]
      ]];
    }

    if(arcgis.compressedGeometry || arcgis.geometry || arcgis.attributes) {
      geojson.type = "Feature";

      if(arcgis.compressedGeometry){
        arcgis.geometry = {
          paths: [
            decompressGeometry(arcgis.compressedGeometry)
          ]
        };
      }

      geojson.geometry = (arcgis.geometry) ? parse(arcgis.geometry) : null;
      geojson.properties = (arcgis.attributes) ? clone(arcgis.attributes) : null;
      if(arcgis.attributes) {
        geojson.id =  arcgis.attributes[options.idAttribute] || arcgis.attributes.OBJECTID || arcgis.attributes.FID;
      }
    }

    return new Terraformer.Primitive(geojson);
  }

  // GeoJSON -> ArcGIS
  function convert(geojson, options){
    var spatialReference;

    options = options || {};
    var idAttribute = options.idAttribute || "OBJECTID";

    if(options.sr){
      spatialReference = { wkid: options.sr };
    } else if (geojson && geojson.crs && geojson.crs.properties.name != "urn:ogc:def:crs:OGC:1.3:CRS84") {
      spatialReference = null;
    } else {
      spatialReference = { wkid: 4326 };
    }

    var result = {};
    var i;

    switch(geojson.type){
    case "Point":
      result.x = geojson.coordinates[0];
      result.y = geojson.coordinates[1];
      if(geojson.coordinates[2]) {
        result.z = geojson.coordinates[2];
      }
      if(geojson.coordinates[3]) {
        result.m = geojson.coordinates[3];
      }
      result.spatialReference = spatialReference;
      break;
    case "MultiPoint":
      result.points = geojson.coordinates.slice(0);
      result.spatialReference = spatialReference;
      break;
    case "LineString":
      result.paths = [geojson.coordinates.slice(0)];
      result.spatialReference = spatialReference;
      break;
    case "MultiLineString":
      result.paths = geojson.coordinates.slice(0);
      result.spatialReference = spatialReference;
      break;
    case "Polygon":
      result.rings = orientRings(geojson.coordinates.slice(0));
      result.spatialReference = spatialReference;
      break;
    case "MultiPolygon":
      result.rings = flattenMultiPolygonRings(geojson.coordinates.slice(0));
      result.spatialReference = spatialReference;
      break;
    case "Feature":
      if(geojson.geometry) {
        result.geometry = convert(geojson.geometry, options);
      }
      result.attributes = (geojson.properties) ? clone(geojson.properties) : {};
      if(geojson.id) {
        result.attributes[idAttribute] = geojson.id;
      }
      break;
    case "FeatureCollection":
      result = [];
      for (i = 0; i < geojson.features.length; i++){
        result.push(convert(geojson.features[i], options));
      }
      break;
    case "GeometryCollection":
      result = [];
      for (i = 0; i < geojson.geometries.length; i++){
        result.push(convert(geojson.geometries[i], options));
      }
      break;
    }

    return result;
  }

  function parseCompressedGeometry(string){
    return new Terraformer.LineString(decompressGeometry(string));
  }

  exports.parse   = parse;
  exports.convert = convert;
  exports.toGeoJSON = parse;
  exports.fromGeoJSON = convert;
  exports.parseCompressedGeometry = parseCompressedGeometry;

  return exports;
}));


terraformer-arcgis-parser.d.ts
declare namespace Terraformer {
    namespace ArcGIS {
        namespace ArcGIS {
            interface SpatialReference {
            }

            interface SpatialReferenceWkid extends SpatialReference {
                wkid?: number;
                latestWkid?: number;
                vcsWkid?: number;
                latestVcsWkid?: number;
            }

            interface SpatialReferenceWkt extends SpatialReference {
                wkt?: string;
                latestWkt?: string;
            }

            interface Geometry {
                spatialReference?: SpatialReference
            }

            interface HasZM {
                hasZ?: boolean;
                hasM?: boolean;
            }

            interface Point extends Geometry {
                x: number;
                y: number;
                z?: number;
                m?: number;
            }

            interface Polyline extends HasZM, Geometry {
                paths: number[][][];
            }

            interface Polygon extends HasZM, Geometry {
                rings: number[][][];
            }

            interface Multipoint extends HasZM, Geometry {
                points: number[][];
            }

            interface Envelope extends Geometry {
                xmin: number;
                xmax: number;
                ymin: number;
                ymax: number;

                zmin?: number;
                zmax?: number;

                mmin?: number;
                mmax?: number;
            }

            interface Feature {
                geometry: Geometry;
                attributes: any;
            }

            interface Field {
                name: string;
                type: string;
                alias?: string;
                length?: number;
            }

            type esriGeometryType = "esriGeometryPoint" | "esriGeometryMultipoint" | "esriGeometryPolyline" | "esriGeometryPolygon" | "esriGeometryEnvelope";

            interface FeatureSet extends HasZM {
                objectIdFieldName?: string; //optional
                globalIdFieldName?: string; //optional
                displayFieldName?: string; //optional
                geometryType?: esriGeometryType; //for feature layers only
                spatialReference?: SpatialReference; //for feature layers only.
                fields?: Field[];
                features: Feature[];
            }
        }



        interface ParseOptions {
            sr?: number;
            idAttribute?: string;
        }
        interface ConvertOptions {
            idAttribute?: string;
        }
        function parse<T extends ArcGIS.Geometry>(json: T, options?: ParseOptions): GeoJSON.GeometryObject;
        function parse<T extends GeoJSON.GeometryObject>(json: ArcGIS.Feature, options?: ParseOptions): GeoJSON.Feature<T>;

        function convert<T extends GeoJSON.GeometryObject>(geoJSON: GeoJSON.FeatureCollection<T>, options?: ConvertOptions): ArcGIS.FeatureSet
        function convert<T extends GeoJSON.GeometryObject>(geoJSON: GeoJSON.Feature<T>, options?: ConvertOptions): ArcGIS.Feature
        function convert<T extends GeoJSON.GeometryObject>(geoJSON: T, options?: ConvertOptions): ArcGIS.Geometry
    }


}

export = Terraformer.ArcGIS;

benchmarks/benchmarks.js
ArcGIS = require('../terraformer-arcgis-parser');

var Benchmark = require('benchmark');

var suite = new Benchmark.Suite();

// add tests
suite

.add('ArcGIS Point to GeoJSON Point', function() {
  ArcGIS.parse({
    "x": -66.796875,
    "y": 20.0390625,
    "spatialReference": {
      "wkid": 4326
    }
  });
})

.add('ArcGIS Polyline to GeoJSON LineString', function() {
  ArcGIS.parse({
    "paths": [
      [ [6.6796875,47.8125],[-65.390625,52.3828125],[-52.3828125,42.5390625] ]
    ],
    "spatialReference": {
      "wkid": 4326
    }
  });
})

.add('ArcGIS Polygon to GeoJSON Polygon', function() {
  ArcGIS.parse({
    "rings": [
      [ [41.8359375,71.015625],[56.953125,33.75],[21.796875,36.5625],[41.8359375,71.015625] ]
    ],
    "spatialReference": {
      "wkid": 4326
    }
  });
})

.add('ArcGIS Multipoint to GeoJSON Multipoint', function(){
  ArcGIS.parse({
    "points":[ [41.8359375,71.015625],[56.953125,33.75],[21.796875,36.5625] ],
    "spatialReference":{
      "wkid":4326
    }
  });
})

.add('ArcGIS Polyline w/ 2 Paths to GeoJSON MultiLineString', function(){
  ArcGIS.parse({
    "paths":[
      [ [41.8359375,71.015625],[56.953125,33.75] ],
      [ [21.796875,36.5625],[41.8359375,71.015625] ]
    ],
    "spatialReference":{
      "wkid":4326
    }
  });
})

.add('ArcGIS Polygon w/ 2 Rings to GeoJSON MultiPolygon', function(){
  ArcGIS.parse({
    "rings":[
      [[-122.63,45.52],[-122.57,45.53],[-122.52,45.50],[-122.49,45.48],[-122.64,45.49],[-122.63,45.52],[-122.63,45.52]],
      [[-83,35],[-74,35],[-74,41],[-83,41],[-83,35]]
    ],
    "spatialReference": {
      "wkid":4326
    }
  });
})

.add('ArcGIS Polygon w/ Holes into GeoJSON MultiPolygon', function(){
  ArcGIS.parse({
    "type":"polygon",
    "rings":[
      [ [-100.74462180954974,39.95017165502381],[-94.50439384003792,39.91647453608879],[-94.41650267263967,34.89313438177965],[-100.78856739324887,34.85708140996771],[-100.74462180954974,39.95017165502381] ],
      [ [-99.68993678392353,39.341088433448896],[-99.68993678392353,38.24507658785885],[-98.67919734199646,37.86444431771113],[-98.06395917020868,38.210554846669694],[-98.06395917020868,39.341088433448896],[-99.68993678392353,39.341088433448896] ],
      [ [-96.83349180978595,37.23732027507514],[-97.31689323047635,35.967330282988534],[-96.5698183075912,35.57512048069255],[-95.42724211456674,36.357601429255965],[-96.83349180978595,37.23732027507514] ],
      [ [-101.4916967324349,38.24507658785885],[-101.44775114873578,36.073960493943744],[-103.95263145328033,36.03843312329154],[-103.68895795108557,38.03770050767439],[-101.4916967324349,38.24507658785885] ]
    ],
    "spatialReference":{
      "wkid":4326
    }
  });
})

.add('ArcGIS Feature to GeoJSON Feature', function(){
  ArcGIS.parse({
    "geometry": {
      "rings": [
        [ [41.8359375,71.015625],[56.953125,33.75],[21.796875,36.5625],[41.8359375,71.015625] ]
      ],
      "spatialReference": {
        "wkid": 4326
      }
    },
    "attributes": {
      "foo": "bar"
    }
  });
})

.add('GeoJSON Point to ArcGIS Point', function() {
  ArcGIS.convert({
    "type": "Point",
    "coordinates": [-58.7109375,47.4609375]
  });
})

.add('GeoJSON LineString to ArcGIS Polyline', function() {
  ArcGIS.convert({
    "type": "LineString",
    "coordinates": [ [21.4453125,-14.0625],[33.3984375,-20.7421875],[38.3203125,-24.609375] ]
  });
})

.add('GeoJSON Polygon to ArcGIS Polygon', function() {
  ArcGIS.convert({
    "type": "Polygon",
    "coordinates": [
      [ [41.8359375,71.015625],[56.953125,33.75],[21.796875,36.5625],[41.8359375,71.015625] ]
    ]
  });
})

.add('GeoJSON Polygon w/ hole to ArcGIS Polygon', function(){
  ArcGIS.convert({
    "type": "Polygon",
    "coordinates": [
      [ [100.0,0.0],[101.0,0.0],[101.0,1.0],[100.0,1.0],[100.0,0.0] ],
      [ [100.2,0.2],[100.8,0.2],[100.8,0.8],[100.2,0.8],[100.2,0.2] ]
    ]
  });
})

.add('GeoJSON Multipoint to ArcGIS Multipoint', function(){
  ArcGIS.convert({
    "type": "MultiPoint",
    "coordinates": [ [41.8359375,71.015625],[56.953125,33.75],[21.796875,36.5625] ]
  });
})

.add('GeoJSON MultiLineString to ArcGIS Polyline', function(){
  ArcGIS.convert({
    "type": "MultiLineString",
    "coordinates": [
      [ [41.8359375,71.015625],[56.953125,33.75] ],
      [ [21.796875,36.5625],[47.8359375,71.015625] ]
    ]
  });
})

.add('GeoJSON MultiPolygon to ArcGIS MultiPolygon', function(){
  ArcGIS.convert({
    "type": "MultiPolygon",
    "coordinates": [
      [
        [ [102.0, 2.0], [103.0, 2.0], [103.0, 3.0], [102.0, 3.0], [102.0, 2.0] ]
      ],
      [
        [ [100.0, 0.0], [101.0, 0.0], [101.0, 1.0], [100.0, 1.0], [100.0, 0.0] ]
      ]
    ]
  });
})

.add('GeoJSON MultiPolygon w/ hole to ArcGIS Polygon', function(){
  ArcGIS.convert({
    "type": "MultiPolygon",
    "coordinates": [
      [
        [ [102.0, 2.0], [103.0, 2.0], [103.0, 3.0], [102.0, 3.0], [102.0, 2.0] ]
      ],
      [
        [ [100.0, 0.0], [101.0, 0.0], [101.0, 1.0], [100.0, 1.0], [100.0, 0.0] ],
        [ [100.2, 0.2], [100.8, 0.2], [100.8, 0.8], [100.2, 0.8], [100.2, 0.2] ]
      ]
    ]
  });
})

.add('GeoJSON Feature to ArcGIS Feature', function(){
  ArcGIS.convert({
    "type":"Feature",
    "geometry": {
      "type": "Polygon",
      "coordinates": [
        [ [41.8359375,71.015625],[56.953125,33.75],[21.796875,36.5625],[41.8359375,71.015625] ]
      ]
    },
    "properties": {
      "foo":"bar"
    }
  });
})

.add('GeoJSON Feature Collection to Array of ArcGIS Features', function(){
  ArcGIS.convert({
    "type": "FeatureCollection",
    "features": [{
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [102.0, 0.5]
      },
      "properties": {
        "prop0": "value0"
      }
    }, {
      "type": "Feature",
      "geometry": {
        "type": "LineString",
        "coordinates": [
          [102.0, 0.0],[103.0, 1.0],[104.0, 0.0],[105.0, 1.0]
        ]
      },
      "properties": {
        "prop0": "value0"
      }
    }, {
      "type": "Feature",
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [ [100.0, 0.0],[101.0, 0.0],[101.0, 1.0],[100.0, 1.0],[100.0, 0.0] ]
        ]
      },
      "properties": {
        "prop0": "value0"
      }
    }]
  });
})

.add('GeoJSON Geometry Collection to Array of ArcGIS Geometries', function(){
  ArcGIS.convert({
    "type" : "GeometryCollection",
    "geometries" : [{
      "type" : "Polygon",
      "coordinates" : [[[-95, 43], [-95, 50], [-90, 50], [-91, 42], [-95, 43]]]
    }, {
      "type" : "LineString",
      "coordinates" : [[-89, 42], [-89, 50], [-80, 50], [-80, 42]]
    }, {
      "type" : "Point",
      "coordinates" : [-94, 46]
    }]
  });
})

.on('cycle', function(event) {
  console.log(String(event.target));
})

// .on('complete', function() {
//   console.log('Fastest is ' + this.filter('fastest').pluck('name'));
// })

.run({ 'async': false });